src/Npgsql/BackendMessages/RowDescriptionMessage.cs

```diff
@@ -35,7 +35,7 @@ internal RowDescriptionMessage(int numFields = 10)
         {
             Count = source.Count;
             _fields = new FieldDescription?[Count];
-            for (var i = 0; i < Count; i++)
+            for (var i = 0; i < _fields.Length; i++)
```

Comment:
To remove more bound checks here:
```c#
var fields = new new FieldDescription[Count];
var sourceFields = _source._fields;
for (var i = 0; i < fields.Length; i++)
    fields[i] = sourceFields[i].Clone();
_fields = fields;
```
JIT is better on removing bound checks on locals instead on (class-) fields.

Reply:
@gfoidl have you verified that the above produces better code (e.g. on sharplab.io)? As far as I know the JIT recognizes Length on arrays specifically, and optimizes that when the array reference isn't modified in the loop body.

Reply:
Not that I'm against these optimizations, but in general RowDescription population shouldn't be considered hot; any perf-sensitive scenario should be using prepared statements, where the row description is populated once and reused. Just in terms of focusing effort etc.

Reply:
Yep verifyed. Didn't change in .NET 6. See [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgMBAOwwBtALoMA+tihQA3Fx6bu9CdIYBZcgAoAlA20dLPPRga4ArvgYBeBgAYN1GzamyFKlCuSoFePlre4QwAZtAMxn4SwZ5JADwMgQB0ADIwkgDmGAAWagwA1GUCptZRjs5lboEyAnJhUTU6AOz2Tm3cAL6WlrqJBqRmFpHcnFM+iXXJfT4dMXEJ+gKLaSGqOXmFJeWV1bNRPfVuyqrNrSsrxN11S4PU/UA==)
(Side note: in the dasm L0012 and L0019 show padding for loop alignment -- nice)

> when the array reference isn't modified in the loop body.

When it's a field the JIT can't assume that it isn't modifyed outside the loop body, so it must (?) be conservative here and emit bound checks. Only with a local reference that assumption holds -> bound checks removed.

Reply:
Thanks for the sharplab code - looks right.

Reply:
Only for [trivial](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgMBAOwwBtALoMA+tihQA3Fx6bu9CdIYBZcgAoAlA20dLPPRga4ArvgYBeBgAYN1GzamyFKlCuSoFePlre4QwAZtAMxn4SwZ5JADwMgQB0ADIwkgDmGAAWagwA1GUCptZRjvgVYVE1OgDs9k6N3AC+lpa6iQakZhaR3JyjPol1yZ0+zTFxCfoCM2khqjl5hSXlldUTUe31ZbM288RtdbM91F1AA) loop body the JIT is able to remove the bound checks.

So is it a missed optimization from the JIT? (I don't know).